src/neurostatslib/basis.py

```diff
@@ -11,315 +11,902 @@
+        xi[0],...,xi[n] : (number of samples, )
+            The input samples.
+
+        Returns
+        -------
+            NDArray
+            The generated basis functions.
+
+        Raises
+        ------
+        NotImplementedError
+            If the subclass does not implement the _evaluate method.
+        """
+        # checks on input and outputs
+        self._check_samples_consistency(*xi)
+        self._check_input_number(xi)
+        self._check_full_model_matrix_size(xi[0].shape[0])
+
+        eval_basis = self._evaluate(*xi)
+
+        # checks on the evaluated basis
+        self._check_enough_samples(eval_basis)  # move to the GLM model
+
+        # check the conditioning
+        conditioning = np.linalg.cond(eval_basis)  # model should do that
+        if np.isinf(conditioning):
+            if any(eval_basis.sum(axis=1) == 0):
+                warnings.warn(
+                    "eval_basis has an empty row. No samples in the input domain "
+                    "of at least one basis function. Try to reduce the number of basis or increase the sample"
+                    "size"
+                )
+            else:
+                warnings.warn(
+                    "Linearly dependent columns in eval_basis. Check for perfect collinearity in the inputs"
+                    "or insufficient sample size."
                 )
-        except AttributeError:
-            raise TypeError("Input is not an array!")
 
-    def _check_num_sample_pts(self, sample_pts: NDArray):
-        """Raise warning if number of sample points looks off.
+        print(f"Conditioning of the evaluated basis function: {conditioning}")
+        return eval_basis
 
-        It's not necessarily *wrong* if ``len(sample_pts) !=
-        self.window_size``, but that's generally what we want, so raise a
-        warning if so.
+    def evaluate_on_grid(self, *n_samples: int) -> Tuple[NDArray, ...]:
+        """
+        Evaluate the basis set on a grid of equi-spaced sample points. The i-th axis of the grid will be sampled
+        with n_samples[i] equi-spaced points.
+
+        Parameters
+        ----------
+        n_samples[0],...,n_samples[n]
+            The number of samples in each axis of the grid.
+
+        Returns
+        -------
+            Xs[1], ..., Xs[n] : (n_samples[0], ... , n_samples[n])
+            A tuple containing the meshgrid values, one element for each of the n dimension of the grid, where n equals
+            to the number of inputs.
+            Y : (number of basis, n_samples[0], ... , n_samples[n])
+            the basis function evaluated at the samples.
 
         """
-        if len(sample_pts) != self.window_size:
+        self._check_input_number(n_samples)
+        n_basis: int = np.prod(n_samples, dtype=int)
+        self._check_full_model_matrix_size(n_basis)
+
+        # get the samples
+        sample_tuple = self._get_samples(*n_samples)
+        Xs = np.meshgrid(*sample_tuple, indexing="ij")
```

Comment:
okay, this all makes sense to me, agree it's much cleaner. still think it's worth mentioning in the `Notes` section of the docstring or in `Returns`. want to prevent someone who's very used to `meshgrid` getting tripped up from the reversal.